### ISSUE TITLE
Sphinx Databus: Implement PDBx/MMCIF

### ISSUE THREAD

**Issue by lizutah** [Electrostatics/APBS_Sphinx#1]
http://mmcif.wwpdb.org/pdbx-mmcif-home-page.htm
- [x] download and test; 
- [x] python source file in sphinx that lets us use that scheme in that code; 
- [x] start creating plugins that use it; 
- [x] define a process for converting all places we pass data into subsets of that schema.

**Comment by keith923** [comment]
For what it's word, the link in the issue description is dead.

My current understanding is that PDBx/mmCIF files are an alternative to PDB files.  The format is [documented](http://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v40.dic/Index/) as a "dictionary", and uses a "funky" format called STAR (which is the same format as the PDBx/mmCIF files).  It is [fully specified here](http://xrpp.iucr.org/cgi-bin/itr?url_ver=Z39.88-2003&rft_dat=what%3Dchapter%26volid%3DGa%26chnumo%3D2o1%26chvers%3Dv0001).  A brief introduction to the syntax is [provided](http://mmcif.wwpdb.org/docs/tutorials/mechanics/pdbx-mmcif-syntax.html).  There is also an interesting read at this [site](http://www.globalphasing.com/startools/#4).

This STAR format is basically a set of key-value pairs with some added fluff for creating tables.  It's very general, in a similar way as JSON or XML, but not as pretty as the former, and less verbose than the latter.  It is this generallity that makes it possible to use the same format to represent both the molecular data, as well as the format for for the molecular data.

But wait, there's more!  The specification for the molecular data, i.e., the PDBx/mmCIF files, is based on an even more abstract [definition](http://mmcif.wwpdb.org/docs/tutorials/mechanics/pdbx-mmcif-dict-struct.html)!  I feel Excedrin headache number 42 coming on!

My thinking is that we'll need to support reading and writing the PDBx/mmCIF format, in addition to using the PDBx/mmCIF specification as the basis for our internal data structure.  Along those lines, it seems reasonable to have code that reads the STAR formatted files into Python data structures that we can export as JSON.  From there we'll be able to either generate a JSON schema for our data structure, or use the JSON as molecular data for the pipeline.

I'll add more specific information as I learn more.

**Comment by keith923** [comment]
This [presentation](http://ipc.cnb.csic.es/download/i2pc-developer-workshop/VocabulariesAndOntologies_JW.pdf) is a good resource for what I was trying to describe above (wish I'd found it first).  Slide 7 is particularly useful in showing how the format is defined.

**Comment by keith923** [comment]
I'm working on converting the XYZR file parser and geoflow plugin to use the JSON schema that I've generated from the PDBx/mmCIF schema.

**Comment by keith923** [comment]
I've got the XYZR and Geometric Flow plugins passing data using what I came up with using the PDBx/mmCIF schema.  It's all currently in the `databus` branch.

In general, there is a new `TypeManager` class that plugins have access to in order to create value instances that are validated against the schema.  There is also some rudimentary support for extending the schema.  I think that it's a good start, and that it's something that we can build upon and refine as necessary.

As for the last "checkbox" at the top of this task...  We are still using ad hoc data formats in the file reader/writer plugins and the stdout plugin.  I think that I'll interpret that task to mean that we need to define something to hold lines of plain text in the schema, and then update the plugins.

As an aside, the entire pipeline source/sink type impedance matching is still defunct.  It's not being used, and it's broken.  I think that we need to specify keys in the schema that correspond to the data that they can source/sink, and then just check those when we create a pipeline.  The question is whether or not we want to do that in this issue, or if there is another place that can be identified as more suitable.  I'll check into that tomorrow, and consult @lizutah.
